# 1. Problem Statement & Business Context

In Singapore's HDB resale market, the buyer and seller must agree on a price before the official HDB valuation is released, so both sides negotiate without knowing what the flat is objectively worth. This hurts both parties: if the agreed price ends up above the valuation, the buyer pays the difference in cash(cash-over-valuation), while sellers risk overpricing and having the flat sit unsold, or underpricing and losing tens of thousands of dollars. This project builds a model that predicts a flat's resale price from its attributes (town, flat type, floor area, storey range, remaining lease) to give both sides an objective starting point. Since the target, resale_price, is a continuous dollar value and not a category, this is a regression problem.

## 2. Data Loading & Initial Inspection

In [1]:
import pandas as pd
df = pd.read_csv("data/resale-flat-prices-2017-onwards.csv")
print(df.shape)

(235247, 11)


In [2]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 235247 entries, 0 to 235246
Data columns (total 11 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   month                235247 non-null  str    
 1   town                 235247 non-null  str    
 2   flat_type            235247 non-null  str    
 3   block                235247 non-null  str    
 4   street_name          235247 non-null  str    
 5   storey_range         235247 non-null  str    
 6   floor_area_sqm       235247 non-null  float64
 7   flat_model           235247 non-null  str    
 8   lease_commence_date  235247 non-null  int64  
 9   remaining_lease      235247 non-null  str    
 10  resale_price         235247 non-null  float64
dtypes: float64(2), int64(1), str(8)
memory usage: 19.7 MB


In [3]:
df.describe()

,floor_area_sqm,lease_commence_date,resale_price
count,235247.000000,235247.000000,2.352470e+05
mean,96.688704,1996.565376,5.320577e+05
std,24.016118,14.367138,1.907679e+05
min,31.000000,1966.000000,1.400000e+05
25%,81.000000,1985.000000,3.900000e+05
50%,93.000000,1997.000000,5.000000e+05
75%,112.000000,2012.000000,6.400000e+05
max,366.700000,2022.000000,1.728000e+06


In [4]:
print("Duplicate rows:", df.duplicated().sum())
print(df.isnull().sum())

Duplicate rows: 316
month                  0
town                   0
flat_type              0
block                  0
street_name            0
storey_range           0
floor_area_sqm         0
flat_model             0
lease_commence_date    0
remaining_lease        0
resale_price           0
dtype: int64


In [5]:
print(df['month'].min(), "to", df['month'].max())
print(df['town'].nunique(), "towns")
print(df['flat_model'].nunique(), "flat models")
print(df['flat_type'].value_counts())
print(df['storey_range'].unique())
df['remaining_lease'].head(10)

2017-01 to 2026-07
26 towns
21 flat models
flat_type
4 ROOM              99899
5 ROOM              57542
3 ROOM              55997
EXECUTIVE           16733
2 ROOM               4902
1 ROOM                 87
MULTI-GENERATION       87
Name: count, dtype: int64
<StringArray>
['10 TO 12', '01 TO 03', '04 TO 06', '07 TO 09', '13 TO 15', '19 TO 21',
 '22 TO 24', '16 TO 18', '34 TO 36', '28 TO 30', '37 TO 39', '49 TO 51',
 '25 TO 27', '40 TO 42', '31 TO 33', '46 TO 48', '43 TO 45']
Length: 17, dtype: str


0    61 years 04 months
1    60 years 07 months
2    62 years 05 months
3     62 years 01 month
4    62 years 05 months
5              63 years
6    61 years 06 months
7    58 years 04 months
8    61 years 06 months
9    61 years 04 months
Name: remaining_lease, dtype: str

## 3. Data Cleaning & Preprocessing

The dataset spans from january 2017 to july 2026. HDB resale prices have shifted alot over this period, especially the post 2020 surge, so training on decade-old prices would distort valuations for today's market. I filter to transactions from Jan 2024 onwards, recent enough to reflect current market conditions, while still keeping a large sample.

In [7]:
print("Before filtering:", df.shape)
df = df[df['month'] >= '2024-01'].copy()
print("After filtering to Jan 2024 onwards:", df.shape)

Before filtering: (66097, 11)
After filtering to Jan 2024 onwards: (66097, 11)


The full dataset contains 316 exact duplicated rows. There is no unit number column so i could not tell whether these are errors or just identical transactions, so i drop. if identical rows appear in the training or test sets, the test score becomes artificially inflated which makes test score look better than it accually is. Although I loses  0.2% of rows, this is still a small cost for a trustworthy evaluation.After filtering to Jan 2024 onwards, no duplicates remained — all 316 were in the older data. I keep this check in as verification that the working dataset is clean.

In [9]:
before = len(df)
df = df.drop_duplicates()
print("Dropped", before - len(df), "duplicates. New shape:", df.shape)

Dropped 0 duplicates. New shape: (66062, 11)


## 4. Exploratory Data Analysis & Interpretation

## 5. Feature Engineering

## 6. Baseline Model (DummyRegressor)

## 7. Model Training & Comparison


## 8. Hyperparameter Tuning (RandomizedSearchCV)

## 9. Final Evaluation & Metric Justification

## 10. Conclusion & Model Export